armasec/token_security.py

```diff
@@ -80,19 +75,21 @@ async def __call__(self, request: Request) -> TokenPayload:
         is injected to a route endpoint via the Depends() method. Lazily loads the OIDC config,
         the TokenDecoder, and the TokenManager if they are not already initialized.
         """
-        if self.manager is None:
-            self.debug_logger("Lazy loading TokenManager")
-            loader = OpenidConfigLoader(self.domain, debug_logger=self.debug_logger)
-            decoder = TokenDecoder(loader.jwks, self.algorithm, debug_logger=self.debug_logger)
-            self.manager = TokenManager(
-                loader.config,
-                decoder,
-                audience=self.audience,
-                debug_logger=self.debug_logger,
-            )
+
+        token_payload = None
 
         try:
-            token_payload = self.manager.extract_token_payload(request.headers)
+            for domain_config in self.domains_config:
+                try:
+                    manager = self._retrieve_manager(domain_config)
```

Comment:
This workflow will make lazy-loading a manager a bit broken for apps that have more than one domain. Basically, every time the app gets a token that doesn't agree with the manager they have lazy loaded, it will cause the manager to be reloaded from scratch.

I think we will need to keep lazy-loaded managers as a mapping between a specific domain and a manager. Then, instead of having to go through the extra work of instantiating a manager only to have it fail to extract the token payload, you could just check to see if you have a manager loaded for that domain yet.

Reply:
if a manager can be loaded than the same manager will be loaded every time. The only slow part is the first call indeed.